### PR TITLE
proxy auth

### PR DESCRIPTION
**Pull request by yangxuanjia**
Grpc-proxy use user and password for access ETCD Server "__lostLeaderKey" when ETCD server enable auth.

If ETCD enable auth, but Grpc-proxy not set user.
leader.go recLoop() will fall in dead Loop for check __lostLeaderKey and response "NOT PERMISSION".

If ETCD cluster all down, and then restart them all(or auto recover).
The Grpc-Proxy watchBoastcast will lose token, watch response "NOT PERMISSION" and never receive etcd watch response event again. Unless we restart Grpc-proxy service.

This PR will fix the bug.

### CODE REVIEW

**Comment by yangxuanjia** [comment]
@ptabor

File: server/etcdmain/grpc_proxy.go [hunk 1]
```diff
@@ -352,6 +362,24 @@ func newClientCfg(lg *zap.Logger, eps []string) (*clientv3.Config, error) {
...
 		lg.Info("gRPC proxy client TLS", zap.String("tls-info", fmt.Sprintf("%+v", tls)))
 	}
+
+	if grpcUser != "" {
+		var err error
+		if grpcPassword == "" {
+			splitted := strings.SplitN(grpcUser, ":", 2)
+			if len(splitted) < 2 {
+				cfg.Username = grpcUser
+				cfg.Password, err = speakeasy.Ask("Password: ")
```
Comment (mitake) [inline server/etcdmain/grpc_proxy.go:372]: Just a quick comment (I'm still reading the change though): getting password in a prompt wouldn't be suitable for server side components. If it's ok for you I suggest to drop this feature. How do you think?

Comment (yangxuanjia) [reply server/etcdmain/grpc_proxy.go:372]: well, I will rid it.  I copy it from etcdctl, haha >_<

Comment (yangxuanjia) [reply server/etcdmain/grpc_proxy.go]: done

File: client/v3/watch.go [hunk 2]
```diff
@@ -581,10 +581,12 @@ func (w *watchGrpcStream) run() {
 			switch {
 			case pbresp.Created:
 				// response to head of queue creation
-				if ws := w.resuming[0]; ws != nil {
-					w.addSubstream(pbresp, ws)
-					w.dispatchEvent(pbresp)
-					w.resuming[0] = nil
+				if len(w.resuming) != 0 {
```
Comment (mitake) [inline client/v3/watch.go:584]: Sorry for being late. Did you see issues related to this conditional branch? I think it is an independent issue of clientv3 (not an issue of gRPC proxy). Could you open a new PR just for submitting the change?

Comment (yangxuanjia) [reply client/v3/watch.go:584]: I hope keep this code, otherwise this will make a panic when etcd server all down and recover, this PR will not complete.

Comment (mitake) [reply client/v3/watch.go:584]: yeah, I meant it's great if you can create an individual PR for this change at first and merge to the master branch. I think the change is important not only for grpc proxy but also for the entire etcd components.

Comment (yangxuanjia) [reply client/v3/watch.go:584]: https://github.com/etcd-io/etcd/pull/12925/files

Comment (mitake) [reply client/v3/watch.go:584]: thanks!

**Comment by mitake** [comment]
@yangxuanjia It seems that https://github.com/etcd-io/etcd/pull/12925 is merged to the master branch, could you rebase this PR on the latest master? Thanks in advance!

**Comment by yangxuanjia** [comment]
> @yangxuanjia It seems that #12925 is merged to the master branch, could you rebase this PR on the latest master? Thanks in advance!

done

File: server/go.mod [hunk 3]
```diff
@@ -3,6 +3,7 @@ module go.etcd.io/etcd/server/v3
 go 1.16
 
 require (
+	github.com/bgentry/speakeasy v0.1.0
```
Comment (mitake) [inline server/go.mod:6]: Could you remove this line? It shouldn't be needed anymore.

File: server/go.sum [hunk 4]
```diff
@@ -30,6 +30,7 @@ github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973/go.mod h1:Dwedo/Wpr24
 github.com/beorn7/perks v1.0.0/go.mod h1:KWe93zE9D1o94FZ5RNwFwVgaQK1VOXiVxmqh+CedLV8=
 github.com/beorn7/perks v1.0.1 h1:VlbKKnNfV8bJzeqoa4cOKqO6bYr3WgKZxO8Z16+hsOM=
 github.com/beorn7/perks v1.0.1/go.mod h1:G2ZrVWU2WbWT9wwq4/hrbKbnv/1ERSJQ0ibhJ6rlkpw=
+github.com/bgentry/speakeasy v0.1.0 h1:ByYyxL9InA1OWqxJqqp2A5pYHUrCiAL6K3J+LKSsQkY=
```
Comment (mitake) [inline server/go.sum:33]: ditto

File: server/go.mod (hunk 3 above)
Comment (mitake) [reply server/go.mod:6]: Do we need the library?

File: server/proxy/grpcproxy/watch_broadcast.go [hunk 5]
```diff
@@ -65,13 +66,46 @@ func newWatchBroadcast(lg *zap.Logger, wp *watchProxy, w *watcher, update func(*
...
 
-		wch := wp.cw.Watch(cctx, w.wr.key, opts...)
-		wp.lg.Debug("watch", zap.String("key", w.wr.key))
-
-		for wr := range wch {
-			wb.bcast(wr)
-			update(wb)
+		var isRetry bool
+
+	OutLayer:
```
Comment (mitake) [inline server/proxy/grpcproxy/watch_broadcast.go:71]: The logic is a little bit complicated so I need a few more days to check, sorry for keeping you waiting.

**Review by yangxuanjia — COMMENTED** [review]
etcdctl use this speakeasy.

**Comment by mitake** [comment]
@yangxuanjia I revisited the design of gRPC proxy + auth, and it assumes that tokens are provided by its clients. So current gRPC proxy is designed not to issue by itself.
If we want to implement the strategy, it will be a little bit hard for the existing mechanism and your design.
Do you have use cases that gRPC proxy should be authorized and clients should use the token of proxy? If you have strong motivation, I'd like to help as much as possible.

**Comment by yangxuanjia** [comment]
> @yangxuanjia I revisited the design of gRPC proxy + auth, and it assumes that tokens are provided by its clients. So current gRPC proxy is designed not to issue by itself.
> If we want to implement the strategy, it will be a little bit hard for the existing mechanism and your design.
> Do you have use cases that gRPC proxy should be authorized and clients should use the token of proxy? If you have strong motivation, I'd like to help as much as possible.

@mitake The current ETCD design allow auth token method. And ETCD-grpc proxy also support auth. But when client connect etcd-grpc-proxy, the first step is invoke authentication() to get token. the put token to ctx, and use the ctx to op.
But the problem is the token not save in proxy. So when token invalid, etcd-grpc-proxy can't recover by itself, it cause the watch failed all the time.
Please think about the situation, or do you have any other suggest?

**Comment by mitake** [comment]
@yangxuanjia I see, let me think about that… it’s a little bit complicated. Probably we should update the logic of watch in grpc proxy.

**Comment by mitake** [comment]
@yangxuanjia sorry for keeping you waiting, the issue is complicated so I'm still thinking about it. BTW is grpc proxy mandatory for your use case?

**Comment by yangxuanjia** [comment]
yes, GRPC-Proxy and user/pass auth method are used in our production env.

**Pull request closed without merging**